src/charm.py

```diff
@@ -1,104 +1,395 @@
 #!/usr/bin/env python3
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
-#
-# Learn more at: https://juju.is/docs/sdk
-
-"""Charm the service.
-
-Refer to the following post for a quick-start guide that will help you
-develop a new k8s charm using the Operator Framework:
-
-    https://discourse.charmhub.io/t/4208
-"""
 
+import json
 import logging
+import os
+from pathlib import Path
+from typing import Union
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+import charms.alertmanager_k8s.v0.alertmanager_remote_configurer as remote_config_write
+import yaml
+from charms.observability_libs.v1.kubernetes_service_patch import (
+    KubernetesServicePatch,
+    ServicePort,
+)
+from ops.charm import (
+    CharmBase,
+    ConfigChangedEvent,
+    PebbleReadyEvent,
+    RelationJoinedEvent,
+)
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    ModelError,
+    WaitingStatus,
+)
+from ops.pebble import Layer
+
+from config_dir_watcher import (
+    AlertmanagerConfigDirWatcher,
+    AlertmanagerConfigFileChangedCharmEvents,
+    AlertmanagerConfigFileChangedEvent,
+)
 
 logger = logging.getLogger(__name__)
 
 
-class OperatorTemplateCharm(CharmBase):
-    """Charm the service."""
+class AlertmanagerConfigurerOperatorCharm(CharmBase):
+    ALERTMANAGER_CONFIG_DIR = "/etc/alertmanager/"
+    ALERTMANAGER_CONFIG_FILE = os.path.join(ALERTMANAGER_CONFIG_DIR, "alertmanager.yml")
+    DUMMY_HTTP_SERVER_SERVICE_NAME = "dummy-http-server"
+    DUMMY_HTTP_SERVER_HOST = "localhost"
+    DUMMY_HTTP_SERVER_PORT = 80
+    ALERTMANAGER_CONFIGURER_SERVICE_NAME = "alertmanager-configurer"
+    ALERTMANAGER_CONFIGURER_PORT = 9101
+    with open(
+        os.path.join(os.path.dirname(os.path.realpath(__file__)), "alertmanager.yml"),
+        "r",
+    ) as default_yaml:
+        ALERTMANAGER_DEFAULT_CONFIG = default_yaml.read()
```

Comment:
Is there any reason why we're using `with...` instead of `yaml.safe_load()`? I thought the second one was more secure, as it prevents Python code from being executed.

Reply:
I'm not sure I understand the question. To use any sort of loading/reading data from a file, you need to open this file first. That's what `with open()` does. `with open()` is in not an alternative to `yaml.safe_load()`.

Reply:
Ok sorry, I meant `.read()`. Do you need to open the file explicitly using either `open` or `with...`? In the integration tests we do `METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())` maybe there's a difference I'm not seeing

Reply:
I'm using `read()` because I need `ALERTMANAGER_DEFAULT_CONFIG` to be a string, while `safe_load()` returns `dict`.

Regarding opening of the file - yes, you can't avoid doing that. If you looked at the implementation of the `read_text()`, you'd see that it's just a wrapper which uses `with open()` and returns `read()`; so exactly the same thing I'm doing here.